### PR TITLE
Move 'datastore.api' functions to 'datastore.client.Client' methods.

### PR DESCRIPTION
**Pull request by tseaver**
Remove the 'datastore.api' module altogether.

See #944.

### CODE REVIEW

File: gcloud/datastore/client.py [hunk 1]
```diff
@@ -27,6 +23,89 @@
...
+_MAX_LOOPS = 128
+"""Maximum number of iterations to wait for deferred keys."""
+
+
+def _extended_lookup(connection, dataset_id, key_pbs,
+                     missing=None, deferred=None,
+                     eventual=False, transaction_id=None):
+    """Repeat lookup until all keys found (unless stop requested).
+
+    Helper function for :method:`Client.get_multi`.
```
Comment (dhermes) [inline gcloud/datastore/client.py]: Double-checked. This is only change from original. (I'm trying to be methodical.)

File: gcloud/datastore/client.py [hunk 2]
```diff
@@ -51,53 +130,168 @@ def __init__(self, dataset_id=None, namespace=None, connection=None):
...
     def get(self, key, missing=None, deferred=None):
-        """Proxy to :func:`gcloud.datastore.api.get`.
+        """Retrieve an entity from a single key (if it exists).
 
-        Passes our ``dataset_id``.
+        .. note::
+
+           This is just a thin wrapper over :method:`get_multi`.
+           The backend API does not make a distinction between a single key or
+           multiple keys in a lookup request.
```
Comment (dhermes) [inline gcloud/datastore/client.py]: I think the linter will complain about undocumented args here, but I will too.

From the old method

```python
    :type key: :class:`gcloud.datastore.key.Key`
    :param key: The key to be retrieved from the datastore.

    :type missing: an empty list or None.
    :param missing: If a list is passed, the key-only entities returned
                    by the backend as "missing" will be copied into it.
                    Use only as a keyword param.

    :type deferred: an empty list or None.
    :param deferred: If a list is passed, the keys returned
                     by the backend as "deferred" will be copied into it.
                     Use only as a keyword param.
```

File: gcloud/datastore/client.py [hunk 3]
```diff
@@ -51,53 +130,168 @@ def __init__(self, dataset_id=None, namespace=None, connection=None):
...
+        :param keys: The keys to be retrieved from the datastore.
+
+        :type missing: an empty list or None.
+        :param missing: If a list is passed, the key-only entities returned
+                        by the backend as "missing" will be copied into it.
+                        Use only as a keyword param.
+
+        :type deferred: an empty list or None.
+        :param deferred: If a list is passed, the keys returned
+                        by the backend as "deferred" will be copied into it.
```
Comment (dhermes) [inline gcloud/datastore/client.py]: looks like indent is off here

File: gcloud/datastore/client.py [hunk 4]
```diff
@@ -51,53 +130,168 @@ def __init__(self, dataset_id=None, namespace=None, connection=None):
...
+                        Use only as a keyword param.
```
Comment (dhermes) [inline gcloud/datastore/client.py]: Where are returns and raises?

File: gcloud/datastore/client.py [hunk 5]
```diff
@@ -51,53 +130,168 @@ def __init__(self, dataset_id=None, namespace=None, connection=None):
...
+                helpers.entity_from_protobuf(missed_pb)
+                for missed_pb in missing]
+
+        if deferred is not None:
+            deferred[:] = [
+                helpers.key_from_protobuf(deferred_pb)
+                for deferred_pb in deferred]
+
+        return [helpers.entity_from_protobuf(entity_pb)
+                for entity_pb in entity_pbs]
```
Comment (dhermes) [inline gcloud/datastore/client.py]: Strange that the thing we used to have do this wasn't a list comprehension. Nice catch.

File: gcloud/datastore/client.py [hunk 6]
```diff
@@ -51,53 +130,168 @@ def __init__(self, dataset_id=None, namespace=None, connection=None):
...
-        return put(entity, connection=self.connection,
-                   dataset_id=self.dataset_id)
+        self.put_multi(entities=[entity])
 
     def put_multi(self, entities):
-        """Proxy to :func:`gcloud.datastore.api.put_multi`.
+        """Save entities in the Cloud Datastore.
 
-        Passes our ``dataset_id``.
+        :type entities: list of :class:`gcloud.datastore.entity.Entity`
```
Comment (dhermes) [inline gcloud/datastore/client.py]: `..note` didn't get ported over

File: gcloud/datastore/client.py [hunk 7]
```diff
@@ -51,53 +130,168 @@ def __init__(self, dataset_id=None, namespace=None, connection=None):
...
+        :param entities: The entities to be saved to the datastore.
         """
```
Comment (dhermes) [inline gcloud/datastore/client.py]: No raises section

File: gcloud/datastore/client.py [hunk 8]
```diff
@@ -51,53 +130,168 @@ def __init__(self, dataset_id=None, namespace=None, connection=None):
...
-        return put_multi(entities, connection=self.connection,
-                         dataset_id=self.dataset_id)
+        if isinstance(entities, Entity):
+            raise ValueError("Pass a sequence of entities")
```
Comment (dhermes) [inline gcloud/datastore/client.py]: It seems we should be more consistent (not in this PR) about how we do this.

File: gcloud/datastore/client.py [hunk 9]
```diff
@@ -51,53 +130,168 @@ def __init__(self, dataset_id=None, namespace=None, connection=None):
...
+        :rtype: list of :class:`gcloud.datastore.key.Key`
+        :returns: The (complete) keys allocated with ``incomplete_key`` as
+                  root.
+        :raises: :class:`ValueError` if ``incomplete_key`` is not a
+                 partial key.
         """
-        return delete_multi(keys, connection=self.connection,
-                            dataset_id=self.dataset_id)
+        if not incomplete_key.is_partial:
+            raise ValueError(('Key is not partial.', incomplete_key))
```
Comment (dhermes) [inline gcloud/datastore/client.py]: When I originally wrote this, I for some reason thought that `ValueError` needed a single argument (hence the `tuple`) in its constructor. We can ditch the tuple and just give 2 args if you like. But it's NBD either way.

File: system_tests/datastore.py [hunk 10]
```diff
@@ -32,16 +32,18 @@ def setUp(self):
         self.case_entities_to_delete = []
 
     def tearDown(self):
-        with datastore.Transaction():
+        client = datastore.Client()
```
Comment (dhermes) [inline system_tests/datastore.py]: Do we want a global client for this module?

**Comment by dhermes** [comment]
I did a full review of everything except `test_client.py`. Any suggestions for how to review?

Also, `_require_dataset_id` and `_require_connection` disappeared from `api.py`. Is this a problem? I suppose `_require_dataset_id` has been covered partially by the constructor of `Client`, but `_require_connection` not really at all (until the `AttributeError` when the user tries to use the connection `None` as a `Connection`).

File: gcloud/datastore/client.py (hunk 2 above)
Comment (tseaver) [inline gcloud/datastore/client.py]: ``tox -r -e lint`` didn't barf.

Addressed via 008e4b5

File: gcloud/datastore/client.py (hunk 3 above)
Comment (tseaver) [inline gcloud/datastore/client.py]: 008e4b5

File: gcloud/datastore/client.py (hunk 4 above)
Comment (tseaver) [inline gcloud/datastore/client.py]: 008e4b5

File: gcloud/datastore/client.py (hunk 7 above)
Comment (tseaver) [inline gcloud/datastore/client.py]: ``get`` doesn't raise anything itself.

Added ``:rtype:`` and ``:returns:`` in 008e4b5

File: system_tests/datastore.py (hunk 10 above)
Comment (tseaver) [inline system_tests/datastore.py]: De gustibus, I guess.  I guess it would speed things up a bit.

**Comment by tseaver** [comment]
> Any suggestions for how to review ``test_client.py``?

Coverage is still 100%, and I moved the tests which didnt exercise "implicit" dataset / connection over mostly wholesale.

> _require_dataset_id and _require_connection disappeared from api.py. Is this a problem?

They were only there to support handling implicit connection / dataset ID for the API methods.  We do exercise creating a ``Client`` instance with no arguments, which should be sufficient AFAICT.

File: gcloud/datastore/client.py (hunk 7 above)
Comment (dhermes) [inline gcloud/datastore/client.py]: In the split view, this is on `put_multi`, but maybe you are seeing something different?

File: gcloud/datastore/client.py [hunk 11]
```diff
@@ -51,53 +130,189 @@ def __init__(self, dataset_id=None, namespace=None, connection=None):
...
+        :raises: ValueError if one or more of ``keys`` has a dataset ID which
+                 does not match our dataset ID.
         """
-        return get_multi(keys, missing=missing, deferred=deferred,
-                         connection=self.connection,
-                         dataset_id=self.dataset_id)
+        if not keys:
+            return []
+
+        ids = list(set([key.dataset_id for key in keys]))
```
Comment (dhermes) [inline gcloud/datastore/client.py]: I just realized the cast to list is not needed, but somewhat irrelevant to this PR.

**Comment by dhermes** [comment]
All right, looks like the last remaining bit is about the raises section of `put_multi`.

I'd like the "It seems we should be more consistent (not in this PR) about how we do this." to be addressed at some point, but not here (i.e. just raising in `_multi` when we don't get a list and maybe vice-versa in the singleton methods)

File: gcloud/datastore/client.py (hunk 7 above)
Comment (tseaver) [inline gcloud/datastore/client.py]: Ugh.  3fc9223 adds it.

**Comment by dhermes** [comment]
Not sure what ugh was directed at but LGTM :smile:

**Pull request merged**